oidc/client.py

```diff
@@ -80,32 +88,38 @@ def verify_and_decode_id_token(self, id_token: str) -> Optional[dict]:
         kid = unverified_header.get("kid")
         jwks = self._retrieve_jwks_related_to_kid(kid)
 
-        if jwks:
-            alg = unverified_header.get("alg")
-            public_key = RSAAlgorithm.from_jwk(json.dumps(jwks))
-            cached_data = {}
-
-            try:
-                decoded_token = jwt.decode(
-                    id_token, public_key, audience=[self.client_id], algorithms=alg
-                )
-
-                if decoded_token.get("nonce"):
-                    # Verify that the cached nonce is present and that
-                    # the provider the nonce was initiated for, is the same
-                    # provider returning it
-                    cached_data = cache.get(decoded_token["nonce"])
-                    if self.cache_nonces and self.auth_server != cached_data.get(
-                        "auth_server"
-                    ):
-                        raise NonceVerificationFailed(
-                            "Failed to verify returned nonce value."
-                        )
+        if not jwks:
+            raise NoJSONWebKeyFound("Failed to retrieve key ID described in Token")
+
+        alg = unverified_header.get("alg")
+        public_key = RSAAlgorithm.from_jwk(json.dumps(jwks))
+        cached_data = {}
+
+        try:
+            decoded_token = jwt.decode(
+                id_token, public_key, audience=[self.client_id], algorithms=alg
+            )
+            if self.cache_nonces:
+                # Verify that the cached nonce is present and that
+                # the provider the nonce was initiated for, is the same
+                # provider returning it
+                nonce = decoded_token.get("nonce")
+                if not nonce:
+                    raise NonceVerificationFailed(
+                        "Failed to verify login request. Missing nonce value"
+                    )
+                cached_data = cache.get(nonce)
+                if not cached_data or self.auth_server != cached_data.get(
+                    "auth_server"
+                ):
+                    raise NonceVerificationFailed(
+                        "Failed to verify returned nonce value"
+                    )
+
                 decoded_token[REDIRECT_AFTER_AUTH] = cached_data.get("redirect_after")
-                return decoded_token
-            except Exception as e:
-                raise e
-        return None
+            return decoded_token
+        except Exception as e:
+            raise e
```

Comment:
I do not think this construct is necessary and probably pylint will flag this as a [broad exception](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/broad-except.html).

The likely exception to handle is `jwt.exceptions.DecodeError`. And that would perhaps require information relevant to the user to take action.

Reply:
Resolved in the latest commit